### ISSUE TITLE
Spec: first end-to-end Clip pipeline (CLI + evaluation harness)

### ISSUE THREAD

**Issue by No-0Peration** [No-0Peration/VoxLens#11]
## Problem Statement

VoxLens can read speech from lip movement, but only in principle. There is no way to point it at a video and get text back. Everything established so far — which checkpoint, which runtime, which extraction stack, what the accuracy is — lives in decision tickets, not in anything runnable.

Worse, the reference implementation VoxLens builds on **does not run on the target machine at all**. Three separate defects abort it before inference, none documented upstream. Anyone starting from the published README on Apple Silicon hits a wall in the first ten minutes.

## Solution

A command-line tool that takes a video Clip and prints a Transcript, with the spans where the Speaker's mouth was unreadable reported as Occlusion rather than silently invented.

```
voxlens VIDEO --checkpoint PATH [--json] [--device hybrid] [--beam 1] [--occlusion-min-frames 3]
```

Human-readable text by default; `--json` emits machine-readable output on stdout so the evaluation harness drives the same path users do. Every default is the measured-best option, so the bare invocation is the correct one.

Alongside it, an evaluation harness that runs the CLI over a corpus and reports WER, so any change to the pipeline can be judged against the baselines already measured rather than by impression.

## User Stories

1. As a developer, I want to point the CLI at a video file and get a Transcript on stdout, so that I can see whether VoxLens works at all without writing Python.
2. As a developer, I want the tool to run on my M4 Pro without patching third-party code by hand, so that setup is not a research project.
3. As a developer, I want a single documented setup path, so that I do not rediscover the environment defects myself.
4. As a developer, I want the tool to fail with a clear message when the video has no readable face, so that I can tell "unusable input" from "broken install".
5. As a developer, I want distinct exit codes for bad input, unreadable input, and model failure, so that a batch script can react differently to each.
6. As a developer, I want `--json` output, so that I can pipe results into other tools without parsing prose.
7. As a developer, I want diagnostics on stderr and only results on stdout, so that `--json` stays pipeable.
8. As a developer, I want the Transcript to name the spans where the mouth was unreadable, so that I know which parts of the video the model could not see.
9. As a developer, I want Occlusion reported as time ranges, so that I can locate the unreadable moment in the source video.
10. As a developer, I want short dropouts of one or two Frames ignored, so that ordinary detector noise is not reported as Occlusion.
11. As a developer, I want the Occlusion threshold configurable, so that I can tune it once I have real footage.
12. As a developer, I want the tool to report how many Frames had no detected face, so that I can tell a marginal video from a good one.
13. As a developer, I want the tool to print its real-time factor, so that I can see whether the streaming goal is still reachable.
14. As a developer, I want to select the execution device, so that I can compare hybrid, GPU, and CPU myself.
15. As a developer, I want hybrid execution by default, so that I get the fastest measured configuration without knowing why.
16. As a developer, I want to set the beam width, so that I can trade accuracy against latency deliberately.
17. As a developer, I want greedy decoding by default, so that the tool is fast enough to iterate with.
18. As a researcher, I want an evaluation harness that runs the CLI over a corpus, so that my measurements test the shipped path rather than a parallel one.
19. As a researcher, I want the harness to report WER, so that my numbers are comparable to published figures.
20. As a researcher, I want to evaluate against both the calibration set and the honest set, so that I am not misled by a benchmark-overfit number.
21. As a researcher, I want the harness to accept a sample stride, so that I can get a fast estimate before committing to a full run.
22. As a researcher, I want per-clip results retained, so that I can inspect the worst failures rather than only the aggregate.
23. As a researcher, I want the harness to report throughput alongside accuracy, so that a change that improves WER while destroying latency is visible.
24. As a maintainer, I want the required third-party patches applied in a documented, reviewable way, so that upgrading the upstream repo does not silently reintroduce them.
25. As a maintainer, I want the MediaPipe version pinned, so that a routine dependency bump does not reintroduce a fatal crash.
26. As a maintainer, I want the checkpoint path supplied by the caller rather than downloaded automatically, so that a 4 GB fetch never happens as a side effect.
27. As a maintainer, I want the tool to state which checkpoint and configuration produced a Transcript, so that results are reproducible.
28. As a maintainer, I want Occlusion detection covered by fast tests, so that its edge cases do not require a 4 GB model to verify.
29. As a maintainer, I want the CLI's JSON shape covered by tests, so that the evaluation harness does not break silently.
30. As a future implementer, I want the code to keep the encoder separable from the decoder, so that Core ML conversion later does not require restructuring.
31. As a future implementer, I want the raw per-Frame detection results preserved, so that word-level Occlusion marking can be added without re-running extraction.

## Implementation Decisions

**Pipeline shape.** Video file → Frames → landmark detection → Mouth Region crops plus a per-Frame detection mask → normalisation → encoder → decoder → Transcript plus Occlusion spans → output rendering. Each stage is a module; the CLI composes them.

**Checkpoint (#3, #10).** USR 2.0 Large from `ahaliassos/usr2`, CC BY-NC 4.0, consistent with VoxLens being non-commercial research. Large rather than Huge: Huge's parameter count is unpublished, and Huge's training data includes LRS2 whose terms are more restrictive. The checkpoint is **supplied by the caller** — never auto-downloaded.

**Runtime (#7).** PyTorch with hybrid execution: encoder on MPS, decoder and beam search on CPU. Measured RTF 0.099 against 0.128 pure-MPS and 0.232 pure-CPU, at identical WER and lower memory. The encoder is dense convolution (suits a GPU); beam search is many tiny sequential ops (suits CPU). `--device` exposes all three. **The encoder must remain separable** — it is the streaming-critical component and the natural Core ML conversion unit.

**Extraction (#9).** MediaPipe FaceLandmarker via the bundled task file and its 478→68 landmark mapping. Measured 4.2 ms/Frame, RTF 0.111, essentially resolution-independent. RetinaFace+FAN is excluded — it requires CUDA.

**Occlusion (#6, as amended).** A Frame is unreadable when the detector returns no landmarks; the rule is cause-agnostic, per `CONTEXT.md`. Three or more consecutive unreadable Frames constitute an Occlusion; runs of one or two are noise and stay interpolated. **Occlusion is reported as time ranges only.** Per-word marking is explicitly not attempted: a prototype established that the CTC head and beam search produce different transcriptions, so CTC spike positions cannot locate beam-search tokens. Reporting spans cannot mislabel Read Text, which is the failure this distinction exists to prevent.

The detection mask must be **captured before interpolation**. The reference preprocessing interpolates across gaps and then asserts every Frame has landmarks, destroying the signal. Interpolation still happens — it keeps the tensor well-formed — but the mask survives alongside it.

Occlusion span computation, extracted from the prototype because it encodes the rule more precisely than prose:

```python
def occlusion_runs(mask, min_frames):
    """mask[i] True == Frame unreadable. Returns runs of >= min_frames."""
    runs, start = [], None
    for i, bad in enumerate(list(mask) + [False]):
        if bad and start is None:
            start = i
        elif not bad and start is not None:
            if i - start >= min_frames:
                runs.append((start, i - 1))
            start = None
    return runs
```

**Adjacent spans must be merged** before rendering — the prototype emitted four consecutive markers in one sentence, which is noise rather than information. The merge distance is a spec-level choice and should itself be a named constant.

**Output contract.** `--json` emits: source path, frame count, fps, duration, the Transcript, Occlusion spans (start/end frame, start/end seconds, duration), the count of undetected Frames, timing (extraction, inference, RTF), and the configuration used (device, beam, threshold, checkpoint identity). Human output prints the Transcript with inline Occlusion markers and a one-line summary on stderr.

**Exit codes.** `0` success · `1` mouth unreadable in too many Frames · `2` input missing or undecodable · `3` model or checkpoint failure. Codes 1 and 2 must stay distinct — one means the video is unusable, the other that the path is wrong.

**Required upstream patches.** Four, all mandatory, applied so they survive an upstream update and are individually reviewable:
1. The vendored espnet CTC prefix scorer hardcodes `cuda`-or-`cpu` and resolves to CPU on MPS, then fails mid-search with an opaque device mismatch. It must take the device from its input tensor.
2. The reference demo selects only `cuda` or `cpu`; it needs an MPS path and the hybrid split.
3. `torchvision.io.read_video` was removed in torchvision 0.28; video loading must not depend on it.
4. **`mediapipe` must be pinned to `0.10.35`.** Version 1.0.1 aborts fatally inside its Metal path — a process kill, not a catchable exception — and forcing the CPU delegate does not help.

**Environment prerequisite.** On a Mac migrated from Intel hardware, the toolchain may be x86_64 throughout and Homebrew's arm64 Python may be broken (its `pyexpat` links against a system libexpat missing a required symbol, which silently empties `platform.mac_ver()` and breaks pip and uv alike). Setup documentation must cover this; it is not obvious from any error message.

**Performance budget.** End-to-end **RTF 0.29** measured on real video — higher than the 0.21 implied by summing component benchmarks, because video decode and tensor assembly sit outside them. Carry 0.29.

## Testing Decisions

Good tests here assert on **external behaviour only**: what the CLI prints, what exit code it returns, what spans it reports. They must not reach into module internals, assert on tensor shapes, or depend on the specific wording of a Transcript — which varies with checkpoint and beam width and is not a stable contract.

**Two seams, deliberately.**

**1. The CLI process boundary — primary.** Tests invoke the CLI and assert on parsed `--json` output. This is the same surface the evaluation harness uses, so tests exercise the shipped path rather than a parallel one. Covered here: JSON shape and required keys; exit code for a missing file; exit code for an undecodable file; exit code for a video with no detectable face; that Occlusion spans appear when the input contains a real gap; that `--occlusion-min-frames` changes what is reported; that stdout stays clean JSON while diagnostics go to stderr; that the reported configuration matches what was requested.

These tests need a **small fixture video** and the checkpoint. The checkpoint is 4 GB and must not enter version control — tests requiring it should skip with a clear message when it is absent, so the suite stays runnable without it.

**2. Occlusion span computation — a pure function.** Frame mask in, spans out. No model, no video, microseconds per case. Covered here: runs shorter than the threshold are ignored; runs at exactly the threshold are reported; runs at the very start and very end of a Clip are handled; adjacent runs merge according to the merge rule; an all-unreadable Clip and an all-readable Clip both behave; an empty mask does not crash.

This second seam exists because these are precisely the cases that are painful to provoke through a fixture video and absurd to verify through a 4 GB model.

**No prior art in this repo** — it contains no implementation yet, so these tests set the conventions rather than follow them.

## Out of Scope

- **Real-time streaming.** The destination per ADR-0001, but this spec is bounded at Clips.
- **Inferred Text.** Occlusion is detected and marked, never filled. The concept stays defined in `CONTEXT.md`.
- **Word-level Occlusion marking.** Requires forced alignment of the beam transcript to encoder Frames; spans only for now.
- **Fine-tuning.** No obtainable unconstrained-English training corpus has been identified; the spec assumes a published checkpoint.
- **Multilingual and multiple simultaneous Speakers.** Exactly one Speaker per Clip.
- **iOS/mobile packaging and Core ML conversion.** Deferred, though the encoder stays separable so conversion remains possible.
- **A graphical or web interface.**
- **Auto-downloading the checkpoint or corpora.**

## Further Notes

**Accuracy expectations.** Measured baselines at beam 1: **34.22% WER** on the full 1,320-clip LRS3 test split, **47.85%** on a 570-clip WildVSR sample. WildVSR is the honest number; LRS3 is a calibration check that has already done its job (24.3% at beam 40 against a published 21.5% confirmed the crop convention is correct). Roughly every other word is wrong on real-world video. That is near the state of the art for video-only lip reading and should be stated plainly rather than discovered later.

**An unresolved discrepancy.** Published figures put this model at 73.7% on WildVSR; this harness measured 47.85% — better, using a smaller configuration, where it should be worse. Most likely a text-normalisation difference. **Until reconciled, the baselines are valid for regression detection against themselves and are not claims of parity with published work.**

**Untested paths.** Both evaluation corpora ship pre-cropped mouth regions, so they exercise the recogniser while bypassing face detection, Mouth Region extraction and Occlusion entirely. The Occlusion path has never run against real footage containing a real gap, and the 3-Frame threshold is reasoned from viseme duration rather than tuned. Detection reliability on genuine video is unmeasured.

**Vocabulary.** Clip, Stream, Frame, Speaker, Mouth Region, Occlusion, Transcript, Read Text, Inferred Text are defined in `CONTEXT.md` and used as defined throughout. Use them in code and output rather than reaching for synonyms.

**Comment by No-0Peration** [comment]
Implemented and shipped. `voxlens` and `voxlens-eval` do what this spec describes, and every user story in it is delivered:

- video in, Transcript on stdout, no Python required
- runs on Apple Silicon without hand-patching third parties — `scripts/vendor.py` applies the three patches
- one documented setup path (`docs/setup.md`), including the migrated-Intel-Mac failure
- clear failure on unreadable input, and distinct exit codes 0/1/2/3 separating bad input from unusable footage from a broken checkpoint
- `--json` on stdout, diagnostics on stderr, so the harness drives the shipped path (ADR-0007)
- Occlusion reported as time ranges beside the Transcript, never spliced into it (ADR-0008)

Two corrections to what this spec asserted, both from measuring rather than reasoning:

**The accuracy figures.** The spec cites 47.85% on WildVSR from a 570-Clip stride sample. The **full 2,854-Clip split scores 50.0%**; the same sample reproduced at 47.78% in the same run, so nothing regressed — sampling was about two points optimistic. The README and the harness now carry the full number.

**One "Out of Scope" line has partly moved.** Word-level Occlusion marking is still out of reach without forced alignment. But the underlying question — how much of this Transcript to trust — now has a measured answer at Clip granularity: divergence between the CTC and beam decoders predicts per-Clip WER at Spearman 0.752, and `--divergence` reports it as a calibrated band ([#22](https://github.com/No-0Peration/VoxLens/issues/22), ADR-0011).

Closing as delivered. What came after it is recorded in [live-capture.md](https://github.com/No-0Peration/VoxLens/blob/main/docs/live-capture.md) and ADRs 0009–0012.
